### PR TITLE
Testing in Go

### PR DESCRIPTION
**Pull request by elisarver**
- support JSON logging globally
 - code to run the makefiles to clean-up
 - code to launch whatever test file is in each topic directory

### CODE REVIEW

**Comment by elisarver** [comment]
WIP to gather feedback from @dpastoor

**Review by dpastoor — COMMENTED** [review]
quick question/comment pass. I wonder if its worth going "all in" on these make targets or should we consider maybe just making these resets native to the test harness? 

The one nice thing about the make is often with how they're designed we can iterate/investigate manually eg just manually running pkgr plan, etc, so having something not in the test harness makes it universally accessible. 

My gut reaction is this is actually a super nice way of keeping the manual vs in during testing very unified and I'm liking the pattern!

File: integration_tests/repo-local/repo-local_test.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+	"github.com/metrumresearchgroup/command"
+)
+
+func TestPlanInstall(t *testing.T) {
+	env := os.Environ()
+
+	// this is where we manipulate the environment
+
+	r, err := command.New(command.WithEnv(env)).Run(context.Background(), "pkgr", "--logjson", "plan")
```
[the remark this answers is not in the document]
Comment (elisarver) [reply integration_tests/repo-local/repo-local_test.go:18]: I would say that is a drone concern. It's also a make target. Either way, the thing responsible for the test should probably just reset state of tests and then run the test, leaving the directory in its tested state at the end.

There are also other situations where the developer might have an older copy on their system to reproduce a bug, in which case they may not want to re-install.

File: integration_tests/integration_helper/targets/targets.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		TarballInstall:       "test-tarball-install-reset",
+		TarballRollback:      "test-tarball-rollback-reset",
+		TildeExpansion:       "test-tilde-expansion-reset",
+		BinariesMac:          "test-binaries-mac-reset",
+		BinariesLinux:        "test-binaries-linux-reset",
+		PathBug:              "test-path-bug-reset",
+		EnvVars:              "test-env-vars-reset",
+	}
+)
+
```
[the remark this answers is not in the document]
Comment (elisarver) [reply integration_tests/integration_helper/targets/targets.go:68]: There are ways to generate these, but that effort may be higher than adding new test cases to the list. I'll give it some thought.

**Comment by elisarver** [comment]
> quick question/comment pass. I wonder if its worth going "all in" on these make targets or should we consider maybe just making these resets native to the test harness?
> 
> The one nice thing about the make is often with how they're designed we can iterate/investigate manually eg just manually running pkgr plan, etc, so having something not in the test harness makes it universally accessible.
> 
> My gut reaction is this is actually a super nice way of keeping the manual vs in during testing very unified and I'm liking the pattern!

I kept the make file for the exact reason you mention above. It does make manual testing easier. Note that the cleanup is also run at the *beginning* of the test, meaning the whole result is waiting if you need to compare. Ex: in your change-list where you demo functionality in this repo, you clean up some files at the end (goldie target files, perhaps?)

I think that you should always clean up before your test, because in the face of failure, leaving the files on disk takes no additional intervention. Nor would doing a final manual check if the files still exist. This is why I put cleanup first instead of last.

**Comment by dpastoor** [comment]
100% agree with cleanup being run at beginning of test to make sure env is fresh.

File: integration_tests/repo-local/repo-local_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dpastoor) [reply integration_tests/repo-local/repo-local_test.go:18]: good points - and that would also solve the problem of figuring out when/how to generate pkgr for each test by leaving it up to the caller. So CI we'll have pkgr build in a prior stage/step, then for manual we have the `make install`

**Review by dpastoor — APPROVED** [review]
great additions - will work to roll them into the broader testing refactor. Thanks!!

**Pull request merged**